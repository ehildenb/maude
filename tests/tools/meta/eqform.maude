load ../../../contrib/tools/meta/eqform.maude

fmod EQFORM-TEST-MODULE is
    sorts Foo Bar Baz Wop Stu Roc .
    -------------------------------
    subsorts Wop < Foo Bar < Baz .
    subsorts Roc < Stu .

    op wop : -> Wop .
    op foo : -> Foo .
    op bar : -> Bar .
    op baz : -> Baz .
    op roc : -> Roc .
    op stu : -> Stu .

    op f : Baz -> Baz .
    -------------------
    eq f(baz) = baz .
    eq f(bar) = baz .
    eq f(foo) = bar .
    eq f(wop) = wop .
endfm

fmod EQFORM-TEST is
  inc EQFORM-CNF .
  inc EQFORM-DNF .
  inc EQFORM-LIST .
  inc EQFORM-SET .
  inc EQFORM-OPERATIONS .

  var T : Term .

  op f : Term -> Term .
  ---------------------
  eq f(T) = 'f[T] .

  ops t1 t2 : -> Term .
  ---------------------
  eq t1 = f(f('B:Baz)) .
  eq t2 = 'bar.Bar .

  ops a b c d e : -> EqLit .
  --------------------------
  eq a = 'foo.Foo ?= 'B:Bar .
  eq b = 'F:Foo != f('bar.Bar) .
  eq c = 'H:Baz ?= f('I:Wop) .
  eq d = f('H:Baz) != f(f('I:Wop)) .
  eq e = 'U:Stu ?= 'W:Roc .

  op f1 : -> Form .
  eq f1 = ~ (a /\ b) \/ ~ ((c \/ d) /\ 'K:Foo ?= 'L:Bar \/ (a /\ ff)) .

  op f2 : -> Form .
  eq f2 = (f1 /\ ~ e) /\ (d \/ tt) .
endfm

--- substitutions
reduce ('_+_['1.Nat, 'X:Nat] ?= '_*_['Y:Nat, 'X:Nat] /\ '_<_['Y:Nat, 'Z:Nat] != 'true.Bool) << ('Y:Nat <- '3.Nat)
    == ('_+_['1.Nat, 'X:Nat] ?= '_*_['3.Nat, 'X:Nat] /\ '_<_['3.Nat, 'Z:Nat] != 'true.Bool) .

reduce ('_+_['1.Nat, 'X:Nat] != '_*_['Y:Nat, 'X:Nat] \/ '_<_['Y:Nat, 'Z:Nat] != 'true.Bool) << ('X:Nat <- '7.Nat)
    == ('_+_['1.Nat, '7.Nat] != '_*_['Y:Nat, '7.Nat] \/ '_<_['Y:Nat, 'Z:Nat] != 'true.Bool) .

reduce f1 << ('F:Foo <- 'F2:Foo ; 'I:Wop <- 'I2:Wop ; 'U:Stu <- 'U2:Stu) .
reduce f2 << ('F:Foo <- 'F2:Foo ; 'I:Wop <- 'I2:Wop ; 'U:Stu <- 'U2:Stu) .

--- sort check
reduce f1 :: Form .
reduce f2 :: Form .

--- testing identities
reduce tt /\ ff .
reduce tt \/ ff .

--- testing idempotency
reduce tt /\ F:Form .
reduce ff \/ F:Form .

--- testing transformations
reduce nnf(f1) .
reduce nnf(f2) .

reduce nef(f1) .
reduce nef(f2) .

reduce cnf(f1) .
reduce cnf(f2) .

reduce dnf(f1) .
reduce dnf(f2) .

--- testing operations
reduce wellFormed(upModule('EQFORM-TEST-MODULE), t1 ?= t2) .
reduce wellFormed(upModule('EQFORM-TEST-MODULE), c) .
reduce wellFormed(upModule('EQFORM-TEST-MODULE), f1) .
reduce wellFormed(upModule('EQFORM-TEST-MODULE), f2) .
reduce wellFormed(upModule('EQFORM-TEST-MODULE), f1 /\ 'T:BadSort ?= '0.Term) .

reduce normalize(upModule('NAT), 's_['s_['0.Nat]] ?= '0.Nat) .

reduce normalize(upModule('EQFORM-TEST-MODULE), t1 ?= t2) .
reduce    reduce(upModule('EQFORM-TEST-MODULE), t1 ?= t2) .
reduce normalize(upModule('EQFORM-TEST-MODULE), f(f(t1)) ?= f(t2)) .
reduce    reduce(upModule('EQFORM-TEST-MODULE), f(f(t1)) ?= f(t2)) .
reduce normalize(upModule('EQFORM-TEST-MODULE), f1) .
reduce    reduce(upModule('EQFORM-TEST-MODULE), f1) .
reduce normalize(upModule('EQFORM-TEST-MODULE), f2) .
reduce    reduce(upModule('EQFORM-TEST-MODULE), f2) .

reduce vars('s_ ['s_['X:Nat]] ?= '0.Nat) .
reduce vars(f1) .
reduce vars(f2) .

reduce toUnifProb(a /\ c /\ e) .
