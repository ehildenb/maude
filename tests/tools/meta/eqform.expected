==========================================
reduce in EQFORM-TEST : ('_*_['Y:Nat,'X:Nat] ?= '_+_['1.Nat,'X:Nat] /\
    'true.Bool != '_<_['Y:Nat,'Z:Nat]) << (
  'Y:Nat <- '3.Nat) == ('_*_['3.Nat,'X:Nat] ?= '_+_['1.Nat,'X:Nat] /\
    'true.Bool != '_<_['3.Nat,'Z:Nat]) .
rewrites: 33
result Bool: true
==========================================
reduce in EQFORM-TEST : ('true.Bool != '_<_['Y:Nat,'Z:Nat] \/ '_*_['Y:Nat,
    'X:Nat] != '_+_['1.Nat,'X:Nat]) << (
  'X:Nat <- '7.Nat) == ('true.Bool != '_<_['Y:Nat,'Z:Nat] \/ '_*_['Y:Nat,
    '7.Nat] != '_+_['1.Nat,'7.Nat]) .
rewrites: 36
result Bool: true
==========================================
reduce in EQFORM-TEST : f1 :: Form .
rewrites: 14
result Bool: true
==========================================
reduce in EQFORM-TEST : f2 :: Form .
rewrites: 21
result Bool: true
==========================================
reduce in EQFORM-TEST : ff .
rewrites: 0
result FalseLit: ff
==========================================
reduce in EQFORM-TEST : tt .
rewrites: 0
result TrueLit: tt
==========================================
reduce in EQFORM-TEST : F:Form .
rewrites: 0
result Form: F:Form
==========================================
reduce in EQFORM-TEST : F:Form .
rewrites: 0
result Form: F:Form
==========================================
reduce in EQFORM-TEST : nnf(f1) .
rewrites: 118
result EqForm: 'F:Foo ?= 'f['bar.Bar] \/ 'B:Bar != 'foo.Foo \/ 'W:Wop !=
    'bar.Bar \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f['I:Wop]
==========================================
reduce in EQFORM-TEST : nnf(f2) .
rewrites: 142
result EqForm: 'U:Stu != 'W:Roc /\ 'F:Foo ?= 'f['bar.Bar] \/ 'B:Bar != 'foo.Foo
    \/ 'W:Wop != 'bar.Bar \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f[
    'I:Wop]
==========================================
reduce in EQFORM-TEST : nef(f1) .
rewrites: 120
result EqForm: 'F:Foo ?= 'f['bar.Bar] \/ 'B:Bar != 'foo.Foo \/ 'W:Wop !=
    'bar.Bar \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f['I:Wop]
==========================================
reduce in EQFORM-TEST : nef(f2) .
rewrites: 144
result EqForm: 'U:Stu != 'W:Roc /\ 'F:Foo ?= 'f['bar.Bar] \/ 'B:Bar != 'foo.Foo
    \/ 'W:Wop != 'bar.Bar \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f[
    'I:Wop]
==========================================
reduce in EQFORM-TEST : cnf(f1) .
rewrites: 635
result EqForm: ('F:Foo ?= 'f['bar.Bar] \/ 'f['H:Baz] ?= 'f['f['I:Wop]] \/
    'B:Bar != 'foo.Foo \/ 'W:Wop != 'bar.Bar) /\ 'F:Foo ?= 'f['bar.Bar] \/
    'B:Bar != 'foo.Foo \/ 'H:Baz != 'f['I:Wop] \/ 'W:Wop != 'bar.Bar
==========================================
reduce in EQFORM-TEST : cnf(f2) .
rewrites: 705
result EqForm: 'U:Stu != 'W:Roc /\ ('F:Foo ?= 'f['bar.Bar] \/ 'f['H:Baz] ?= 'f[
    'f['I:Wop]] \/ 'B:Bar != 'foo.Foo \/ 'W:Wop != 'bar.Bar) /\ 'F:Foo ?= 'f[
    'bar.Bar] \/ 'B:Bar != 'foo.Foo \/ 'H:Baz != 'f['I:Wop] \/ 'W:Wop !=
    'bar.Bar
==========================================
reduce in EQFORM-TEST : dnf(f1) .
rewrites: 245
result EqForm: 'F:Foo ?= 'f['bar.Bar] \/ 'B:Bar != 'foo.Foo \/ 'W:Wop !=
    'bar.Bar \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f['I:Wop]
==========================================
reduce in EQFORM-TEST : dnf(f2) .
rewrites: 553
result EqForm: ('F:Foo ?= 'f['bar.Bar] /\ 'U:Stu != 'W:Roc) \/ ('B:Bar !=
    'foo.Foo /\ 'U:Stu != 'W:Roc) \/ ('U:Stu != 'W:Roc /\ 'W:Wop != 'bar.Bar)
    \/ 'f['H:Baz] ?= 'f['f['I:Wop]] /\ 'H:Baz != 'f['I:Wop] /\ 'U:Stu != 'W:Roc
==========================================
reduce in EQFORM-TEST : wellFormed(upModule('EQFORM-TEST-MODULE), f1) .
rewrites: 127
result Bool: true
==========================================
reduce in EQFORM-TEST : wellFormed(upModule('EQFORM-TEST-MODULE), f2) .
rewrites: 159
result Bool: true
==========================================
reduce in EQFORM-TEST : wellFormed(upModule('EQFORM-TEST-MODULE), f1 /\ '0.Term
    ?= 'T:BadSort) .
rewrites: 30
result Bool: false
==========================================
reduce in EQFORM-TEST : normalize(upModule('NAT), '0.Nat ?= 's_['s_['0.Nat]]) .
rewrites: 13
result PosEqLit: '0.Zero ?= 's_^2['0.Zero]
==========================================
reduce in EQFORM-TEST : normalize(upModule('EQFORM-TEST-MODULE), f1) .
rewrites: 105
result Form: ~ ('B:Bar ?= 'foo.Foo /\ 'F:Foo != 'f['bar.Bar]) \/ ~ ('W:Wop ?=
    'bar.Bar /\ 'H:Baz ?= 'f['I:Wop] \/ 'f['H:Baz] != 'f['f['I:Wop]])
==========================================
reduce in EQFORM-TEST : normalize(upModule('EQFORM-TEST-MODULE), f2) .
rewrites: 134
result Form: ~ 'U:Stu ?= 'W:Roc /\ ~ ('B:Bar ?= 'foo.Foo /\ 'F:Foo != 'f[
    'bar.Bar]) \/ ~ ('W:Wop ?= 'bar.Bar /\ 'H:Baz ?= 'f['I:Wop] \/ 'f['H:Baz]
    != 'f['f['I:Wop]])
==========================================
reduce in EQFORM-TEST : vars('0.Nat ?= 's_['s_['X:Nat]]) .
rewrites: 11
result Variable: 'X:Nat
==========================================
reduce in EQFORM-TEST : vars(f1) .
rewrites: 100
result NeQidSet: 'B:Bar ; 'F:Foo ; 'H:Baz ; 'I:Wop ; 'W:Wop
==========================================
reduce in EQFORM-TEST : vars(f2) .
rewrites: 127
result NeQidSet: 'B:Bar ; 'F:Foo ; 'H:Baz ; 'I:Wop ; 'U:Stu ; 'W:Roc ; 'W:Wop
==========================================
reduce in EQFORM-TEST : toUnifProb(a /\ c /\ e) .
rewrites: 7
result UnificationProblem: 'B:Bar =? 'foo.Foo /\ 'H:Baz =? 'f['I:Wop] /\ 'U:Stu
    =? 'W:Roc
Bye.
